spec/index.md

```diff
@@ -1401,78 +1407,89 @@ The 32-bit stable memory System API (`ic0.stable_size`, `ic0.stable_grow`, `ic0.
+    ic0.canister_status : () -> i32;                                                      // *
+    ic0.canister_version : () -> i64;                                                     // *
+
+    ic0.msg_method_name_size : () -> I;                                                   // F
+    ic0.msg_method_name_copy : (dst : I, offset : I, size : I) -> ();                     // F
+    ic0.accept_message : () -> ();                                                        // F
+
+    ic0.call_new :
+      ( callee_src  : I,
+        callee_size : I,
+        name_src    : I,
+        name_size   : I,
+        reply_fun   : I,
+        reply_env   : I,
+        reject_fun  : I,
+        reject_env  : I
+      ) -> ();                                                                            // U CQ Ry Rt CRy CRt T
+    ic0.call_on_cleanup : (fun : I, env : I) -> ();                                       // U CQ Ry Rt CRy CRt T
+    ic0.call_data_append : (src : I, size : I) -> ();                                     // U CQ Ry Rt CRy CRt T
+    ic0.call_cycles_add128 : (amount_high : i64, amount_low: i64) -> ();                  // U Ry Rt T
+    ic0.call_perform : () -> ( err_code : i32 );                                          // U CQ Ry Rt CRy CRt T
+
+    ic0.stable64_size : () -> (page_count : i64);                                         // * s
+    ic0.stable64_grow : (new_pages : i64) -> (old_page_count : i64);                      // * s
+    ic0.stable64_write : (offset : i64, src : i64, size : i64) -> ();                     // * s
+    ic0.stable64_read : (dst : i64, offset : i64, size : i64) -> ();                      // * s
+
+    ic0.certified_data_set : (src : I, size : I) -> ();                                   // I G U Ry Rt T
+    ic0.data_certificate_present : () -> i32;                                             // *
+    ic0.data_certificate_size : () -> I;                                                  // NRQ CQ
+    ic0.data_certificate_copy : (dst : I, offset : I, size : I) -> ();                    // NRQ CQ
+
+    ic0.time : () -> (timestamp : i64);                                                   // *
+    ic0.global_timer_set : (timestamp : i64) -> i64;                                      // I G U Ry Rt C T
+    ic0.performance_counter : (counter_type : i32) -> (counter : i64);                    // * s
+    ic0.is_controller: (src : I, size : I) -> ( result: i32);                             // * s
+    ic0.in_replicated_execution: () -> (result: i32);                                     // * s
+
+    ic0.debug_print : (src : I, size : I) -> ();                                          // * s
+    ic0.trap : (src : I, size : I) -> ();                                                 // * s
+```
+
+The following System API functions are only available if `I = i32`, i.e., if the bit-width of the declared memory is 32
+or if the canister declares no memory.
```

Comment:
Is this really the case in the replica?

Reply:
@adambratschikaye Could you please confirm?

Reply:
Yes, the System APIs use i32 as a default, and switch to i64 if the module declares any memory and it is 64-bit. It's scattered across the code, but you can see an example in `main_memory_type()`.

Reply:
On the other hand, [here](https://github.com/dfinity/ic/blob/7f4cfdf6e469c37c14fc9cd8bc4a5a61cb7ed759/rs/embedders/src/wasmtime_embedder/system_api.rs#L302) it doesn't seem to make difference, e.g., the 64-bit cycles API is supposed to be disabled for 64-bit memories, but [here](https://github.com/dfinity/ic/blob/7f4cfdf6e469c37c14fc9cd8bc4a5a61cb7ed759/rs/embedders/src/wasmtime_embedder/system_api.rs#L967) it seems to be enabled.

Reply:
Ah, I see. You're asking to verify that the 32/64-bit APIs in replica align with the specification. I did a similar check while reviewing some older MRs, but I agree we should do another pass now. cc @alexandru-uta

Reply:
thanks, will have another look.

Reply:
@mraszyk and @berestovskyy , I think these are checked in [validation](https://github.com/dfinity/ic/blob/7f4cfdf6e469c37c14fc9cd8bc4a5a61cb7ed759/rs/embedders/src/wasm_utils/validation.rs#L773) and there's a difference on what imports are allowed in 32 or 64 bit modes.